### PR TITLE
feat(dashboard): confirm overwrite to prevent unintended changes

### PR DESCRIPTION
**Pull request by justinpark**
### SUMMARY
At Airbnb, there's reasonable amount of reporting about unintended overwritten of filter scopes settings.
We believe there must be unknown overwriting on json_metadata by other authors during editing. (json_metadata can be easily overwritten since it passes all attributes including unchanged values.)
This commit introduces the inspection step for this high volatility value changes and then proceed to update with the user's confirmation. (the inspection items are defined [here](https://github.com/apache/superset/pull/21819/files#diff-c070f1a491b40805e52139c971994f682921d2588cf1f06685d30ccf200cd419R44))
This commit also adds the logging for the user confirmation.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/1392866/195903661-131177c7-812d-4a38-8d56-b6e429b27cbd.mov

### TESTING INSTRUCTIONS
- set `ENABLE_CONFIRM_OVERWRITE_DASHBOARD_METADATA` feature flag to true
- go to any dashboard page
- update css or/and filter scopes
- click save to confirm the overwrite values

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [x] Required feature flags: `ENABLE_CONFIRM_OVERWRITE_DASHBOARD_METADATA`
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

cc: @ktmud @john-bodley

### CODE REVIEW

File: superset-frontend/spec/fixtures/mockDashboardState.js [hunk 1]
```diff
@@ -30,3 +31,34 @@ export default {
...
+      newValue:
+        ".navbar {\n    transition: opacity 0.5s ease;\n    opacity: 0.05;\n}\n.navbar:hover {\n    opacity: 1;\n}\n.chart-header .header{\n    font-weight: @font-weight-normal;\n    font-size: 12px;\n}\n/*\nvar bnbColors = [\n    //rausch    hackb      kazan      babu      lima        beach     tirol\n    '#ff5a5f', '#7b0051', '#007A87', '#00d1c1', '#8ce071', '#ffb400', '#b4a76c',\n    '#ff8083', '#cc0086', '#00a1b3', '#00ffeb', '#bbedab', '#ffd266', '#cbc29a',\n    '#ff3339', '#ff1ab1', '#005c66', '#00b3a5', '#55d12e', '#b37e00', '#988b4e',\n ];\n*/\n",
+    },
+    {
+      keyPath: 'json_metadata.filter_scopes',
+      oldValue:
+        '{\n  "122": {\n    "ethnic_minority": {\n      "scope": [\n        "ROOT_ID"\n      ],\n      "immune": []\n    },\n    "gender": {\n      "scope": [\n        "ROOT_ID"\n      ],\n      "immune": []\n    },\n    "developer_type": {\n      "scope": [\n        "ROOT_ID"\n      ],\n      "immune": []\n    },\n    "lang_at_home": {\n      "scope": [\n        "ROOT_ID"\n      ],\n      "immune": []\n    },\n    "country_live": {\n      "scope": [\n        "ROOT_ID"\n      ],\n      "immune": []\n    }\n  }\n}',
```
Comment (john-bodley) [inline superset-frontend/spec/fixtures/mockDashboardState.js:48]: Should we add `\t` here and/or compare the old/new without the formatting, i.e., as a parsed JSON blob?

**Comment by michael-s-molina** [comment]
Thanks for this exciting feature @justinpark! Some first-pass comments:

1 - Can we rename the feature flag to a shorter name like `DASHBOARD_METADATA_DIFF` or anything else?

2 - Can you add the new feature flag to `RESOURCES/FEATURE_FLAGS.md`?

3 - Can we set a min/max width/height to the modal to avoid the situation where it takes the whole screen?

4 - Can we make the modal appear only when there's a change to the scope or CSS? Unrelated changes shouldn't present it.

5 - Is it possible to make the scope changes more user-friendly? It's showing ID diffs that are not traceable to the screen. Maybe add the chart names to the diff or use a different kind of scope representation when comparing.

File: superset-frontend/src/dashboard/actions/dashboardState.js [hunk 2]
```diff
@@ -340,34 +355,87 @@ export function saveDashboardRequest(data, id, saveType) {
...
+            FeatureFlag.ENABLE_CONFIRM_OVERWRITE_DASHBOARD_METADATA,
+          ) ||
+          saveType === SAVE_TYPE_OVERWRITE_CONFIRMED
+        ) {
+          // skip overwrite precheck
+          resolve();
+          return;
+        }
 
-      return SupersetClient.put({
```
Comment (ktmud) [inline superset-frontend/src/dashboard/actions/dashboardState.js:364]: If this dispatch function returns a promise, can we just rewrite it with `async/await`?

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            onClick={onConfirmOverwrite}
+            disabled={!hasReviewed}
+          >
+            {t('Yes, overwrite changes')}
+          </Button>
+        </>
+      }
+      onHide={onHide}
+    >
+      <>
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx:159]: Is this fragment needed?

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+  return (
+    <Modal
+      width="auto"
+      height="50vh"
+      show={Boolean(overwriteConfirmMetadata)}
+      title={t('Confirm overwrite')}
+      footer={
+        <>
+          {t('* Scroll down to the bottom to complete the review. ')}
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx:134]:
```suggestion
          {t('Scroll down to the bottom to complete the review. ')}
```

Not sure if `*` is needed.

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            {t('No')}
+          </Button>
+          <Button
+            data-test="overwrite-confirm-save-button"
+            htmlType="button"
+            buttonSize="small"
+            cta
+            buttonStyle="primary"
+            onClick={onConfirmOverwrite}
+            disabled={!hasReviewed}
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx:151]: Add a tooltip to explain the disabled state?

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+       font-size: 11px;
+       padding: 0px;
+       background-color: transparent;
+       border: 0px;
+       line-height: 110%;
+     }
+   `}
+`;
+
+const StackableHeader = styled(Button)<{ top: number }>`
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx:58]: This header is currently rendered as all caps (probably inherited style from `Button`). Is that what we want?

**Review by ktmud — COMMENTED** [review]
++ to @michael-s-molina 's comments.

A diff modal is a useful add when some of the updates may be implicit but I agree more visual diffs would be nice. I'm also wondering if in the long run we should strive for not requiring diff confirmation as we get more confident that the updates would not accidentally make unwanted changes.

If this is only meant to be temporary---which I'd assume is the case since it's in a feature flag, I think it'd be okay to just show JSON diffs.

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx (hunk 5 above)
Comment (justinpark) [reply superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx:151]: The following description should be enough to explain the disabled state.

my_new_dashboard

**Comment by justinpark** [comment]
> 1 - Can we rename the feature flag to a shorter name like `DASHBOARD_METADATA_DIFF` or anything else?

I updated to `CONFIRM_DASHBOARD_DIFF`

> 2 - Can you add the new feature flag to `RESOURCES/FEATURE_FLAGS.md`?

Yes. updated.

> 3 - Can we set a min/max width/height to the modal to avoid the situation where it takes the whole screen?

Yes I added the maxWidth using `responsive`. (its size up to 1024px)

> 4 - Can we make the modal appear only when there's a change to the scope or CSS? Unrelated changes shouldn't present it.

Sure. The modal only pops up when [the following values](https://github.com/apache/superset/pull/21819/files#diff-c070f1a491b40805e52139c971994f682921d2588cf1f06685d30ccf200cd419R44) are changed.

`export const OVERWRITE_INSPECT_FIELDS = ['css', 'json_metadata.filter_scopes'];`

> 5 - Is it possible to make the scope changes more user-friendly? It's showing ID diffs that are not traceable to the screen. Maybe add the chart names to the diff or use a different kind of scope representation when comparing.

as @ktmud commented, it becomes complicated to access the scope labels and its hierarchy. I'll follow up this improvement in the next step.

File: superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx (hunk 3 above)
Comment (justinpark) [reply superset-frontend/src/dashboard/components/OverwriteConfirm/OverwriteConfirmModal.tsx]: good catch. removed

File: superset-frontend/src/dashboard/actions/dashboardState.js (hunk 2 above)
Comment (justinpark) [reply superset-frontend/src/dashboard/actions/dashboardState.js:364]: It should be better to have a separate PR for the `async/await` migration.

**Comment by justinpark** [comment]
> > 1 - Can we rename the feature flag to a shorter name like `DASHBOARD_METADATA_DIFF` or anything else?
> 
> I updated to `CONFIRM_DASHBOARD_DIFF`
> 
> > 2 - Can you add the new feature flag to `RESOURCES/FEATURE_FLAGS.md`?
> 
> Yes. updated.
> 
> > 3 - Can we set a min/max width/height to the modal to avoid the situation where it takes the whole screen?
> 
> Yes I added the maxWidth using `responsive`. (its size up to 1024px)
> 
> > 4 - Can we make the modal appear only when there's a change to the scope or CSS? Unrelated changes shouldn't present it.
> 
> Sure. The modal only pops up when [the following values](https://github.com/apache/superset/pull/21819/files#diff-c070f1a491b40805e52139c971994f682921d2588cf1f06685d30ccf200cd419R44) are changed.
> 
> `export const OVERWRITE_INSPECT_FIELDS = ['css', 'json_metadata.filter_scopes'];`
> 
> > 5 - Is it possible to make the scope changes more user-friendly? It's showing ID diffs that are not traceable to the screen. Maybe add the chart names to the diff or use a different kind of scope representation when comparing.
> 
> as ktmud commented, it becomes complicated to access the scope labels and its hierarchy. I'll follow up this improvement in the next step.

@michael-s-molina do you have any objections with the approach? We'd like to kick off the feature with JSON diffs first. (and then will follow up the improved labels version)

**Comment by michael-s-molina** [comment]
> @michael-s-molina do you have any objections with the approach? We'd like to kick off the feature with JSON diffs first. (and then will follow up the improved labels version)

No problem 👍🏼

**Comment by justinpark** [comment]
> > @michael-s-molina do you have any objections with the approach? We'd like to kick off the feature with JSON diffs first. (and then will follow up the improved labels version)
> 
> No problem 👍🏼

Great! can you approve the change? (it's blocked by the reviewers)

**Review by ktmud — APPROVED** [review]

**Pull request merged**